clarity/src/vm/analysis/type_checker/natives/mod.rs

```diff
@@ -403,36 +417,81 @@ fn check_contract_call(
             func_signature
         }
         SymbolicExpressionType::Atom(trait_instance) => {
-            // Dynamic dispatch
-            let trait_id = match context.lookup_trait_reference_type(trait_instance) {
-                Some(trait_id) => trait_id,
+            match checker.contract_context.get_variable_type(trait_instance) {
```

Comment:
Given that you have two gates in this `match` statement, I wonder if it would be safer to put the Clarity1-specific code and Clarity2-specific code into different methods that would be called here?  Then, there would be only one gate.

Reply:
I thought this one is pretty clear as-is, with two cases of the match Clarity2-only, and the third (default case) for both versions, but will refactor if you think it makes the code clearer.

Reply:
6b9ee54bd